### DISCUSSION TITLE
How do I start contributing?

### DISCUSSION THREAD

**Discussion by kowshik1234** [micropython/micropython#13191] [Core Development] [answered]
Hi,

I'd like to contribute Micropython code. I am an embedded systems engineer for my entire career and worked for organizations like Texas Instruments, Microchip Technology developing drivers. I am interested in Micropython and would like to contribute, I would like to know below

1. Should I need to have any base board that I need to order to test and contribute?
2. Is there any How-to-contribute guide available somewhere?
3. Any other guidelines for contribution?

Thanks,
Kowshik

**Comment by rkompass** [comment, marked as answer]
One way would be to read the contributions to this discussion for a while, and perhaps back into the past, and decide which of the discussed open problems is the most interesting for you.

A little collection of boards will be helpful: Pyboard 1.1 or D or Blackpill (stm32f411) for the stm side, RP2040 Pico and esp32 appear to be the most used and are not very expensive.

There are many directions to go: Among others drivers (like LCD displays, ADCs, H-bridges) networking, robot stuff. ...

It makes sense to get somewhat acquainted with the GitHub repos of Micropython and Micropython-lib and with Git in general.

There are docs for the python side of Micropython programming. If you want to contribute to the internals you have to get an understanding of the code base mostly yourself. At least there is no systematic documentation of that available.

**Comment by cwalther** [comment]
The contribution guide is right at the top of the repository: https://github.com/micropython/micropython/blob/master/CONTRIBUTING.md

Also, consider contributing by reviewing pull requests. My impression from the number of open PRs and the time it takes to get any response on a PR is that there are too many people contributing code and too few people reviewing those contributions (guilty of it myself).

**Comment by jimmo** [comment]
@kowshik1234

1. It depends what you want to work on. If your interest is in the MicroPython VM and runtime, then almost all development of that happens in the unix port (which runs on e.g. linux/mac/windows). It is useful to test changes on embedded hardware (or at the very least build it for embedded hardware so you can see code size changes), but not strictly necessary.
But if you want to work on drivers or other features specific to a given chip or hardware then you will definitely need that hardware to test on.

2. See the message above linking to the contributing guide.

3. If you're new to git/github you might want to see https://github.com/jimmo/git-and-micropython
